docs/adrs/003-Retry_endpoint_design.md

```diff
@@ -0,0 +1,162 @@
+>I want to see the estimate for a quick and dirty solution that implements a proposed retry endpoint, that can be deployed ASAP, and then once the API >contract has been established, and the functional MVP is done, we begin the cleanup of the backend code.
+>Note that this is NOT business as usual, the usual way we do this is the nasty MVP gets deployed and then we don't go back until much later.
+>Here, we get the API working so that it doesn't block dependencies, and we immediately start the refactoring. The refactor needs to be broken down into >smallish chunks of ~3 days estimated work, and each merge should maintain functionality and incrementally improve the codebase.
+>Tasks that take more than a couple of days are more likely to be far off in their estimate and this is how we mitigate the risk of poor estimation.
+>
+
+
+### High Level Behavior of the `retry` endpoint
+The current implementation of retry is to run jobs using the `retry_job` or `retry_jobs` endpoint. 
+The endpoint takes a job or list of job ids and then attempts to resubmit them to the queue, using the exact same set of parameters and version of the app.
+
+### Current Behavior
+* Spec file is located at https://github.com/kbase/execution_engine2/blob/8baab8e3ac5212f4bbe59fd935980aa41b4ee06d/execution_engine2.spec#L201-L247
+  
+* A job id is provided. If there are sufficient permissions, the call will proceed, if not, it will error out, unless the `as_admin` flag is provided by an admin
+* The retry will only continue if the status of the job to be retried is in [Status.terminated.value, Status.error.value]
+* If the job id points to a job that has already been retried, it will attempt to retry that job's `retry_parent` instead.
+* If the job id has never been retried, it becomes the `retry_parent` 
+* EE2 looks up the method versions and parameters, and then submits the job to be retried, incrementing the `retry_count` of the job being retried, and the newly launched job gains a field called `retry_parent` that contains the job id of the job from the original request.
+* The job is submitted and upon successful submission, the child job adds the field `retry_parent` and notifies the `parent_job_id` that a new `child_job` has been added by appending itself to the `parent_job.child_jobs[]` field
+* There is no way to specify ResourceRequirements with a retry at the moment, even if the job was previously submitted by an administrator and had specfified ResourceRequirements. The retry will only use resource requirements from the catalog / ee2 config.
+
+
+### Batch Behavior
+* If a job has the attribute of `batch_job=True` the retry will fail, since there is no method to re-run. This is a bug, as it doesn't fail gracefully.
+* If a job has the attribute of `batch_job=True`, but is actually a child job, the parent will be notified of this new retried job
```

Comment:
I thought that child jobs were always false, and looking at the current code that seems to be correct unless I'm missing something - how would child jobs wind up with `batch_job=True`?

Reply:
This section talks about the retry endpoint behavior as it was at the time of being written, not the behavior today.

Reply:
Fair enough